### PR TITLE
[test] Set up Vagrant for to provide a consistent OpenLDAP test server

### PR DESCRIPTION
**Pull request by Maks3w**
This is based in the setup original setup instructions https://github.com/zendframework/zf2/blob/eb6275e9227e9ed1995d326087488d3ea25811b5/tests/docs/README-LDAP.txt
the work of pwdng for the FreeRADIUS project FreeRADIUS/freeradius-server#1107
and the first implementation stub of heiglandreas for zend-ldap #8

Supersede #8

### CODE REVIEW

**Comment by Maks3w** [comment]
@weierophinney @ralphschindler Thoughts?

**Comment by weierophinney** [comment]
@Maks3w This looks great! Only thing is: while I succesfully did the `vagrant up`, I'm not sure what I have to do to enable the online tests. Can you document that in the README and/or CONTRIBUTING docs, please?

**Comment by heiglandreas** [comment]
Looks good in so far that it sets up a vagrant box for testing against. 

The tests now have to be adapted. Also there should be a note in the README_TEST.md (which will be included in PR #7 ) that states that the TESTS_ZEND_LDAP_ONLINE_ENABLED has to be set to ```true``` before calling phpunit. ( @weierophinney - that's the trick, we'll document it)

File: phpunit.xml.dist [hunk 1]
```diff
@@ -64,20 +64,20 @@
...
         <env name="TESTS_ZEND_LDAP_USE_SSL" value="false" />
-        <env name="TESTS_ZEND_LDAP_USERNAME" value="CN=someUser,DC=example,DC=com" />
-        <env name="TESTS_ZEND_LDAP_PRINCIPAL_NAME" value="someUser@example.com" />
-        <env name="TESTS_ZEND_LDAP_PASSWORD" value="" />
+        <env name="TESTS_ZEND_LDAP_USERNAME" value="cn=Manager,dc=example,dc=com" />
+        <env name="TESTS_ZEND_LDAP_PASSWORD" value="insecure" />
         <env name="TESTS_ZEND_LDAP_BIND_REQUIRES_DN" value="true" />
-        <env name="TESTS_ZEND_LDAP_BASE_DN" value="OU=Sales,DC=example,DC=com" />
-        <env name="TESTS_ZEND_LDAP_ACCOUNT_FILTER_FORMAT" value="(&amp;(objectClass=posixAccount)(uid=%s))" />
+        <env name="TESTS_ZEND_LDAP_BASE_DN" value="dc=example,dc=com" />
```
Comment (heiglandreas) [inline phpunit.xml.dist]: Why did you change the default values? Changing these means we'll have to adapt all the tests that will fail due to these values not matching the expected ones. (having the attribute-names in lower-case makes actually sense!)

File: phpunit.xml.dist [hunk 2]
```diff
@@ -64,20 +64,20 @@
...
+        <env name="TESTS_ZEND_LDAP_ACCOUNT_FILTER_FORMAT" value="(&amp;(objectClass=account)(uid=%s))" />
         <env name="TESTS_ZEND_LDAP_ACCOUNT_DOMAIN_NAME" value="example.com" />
         <env name="TESTS_ZEND_LDAP_ACCOUNT_DOMAIN_NAME_SHORT" value="EXAMPLE" />
-        <env name="TESTS_ZEND_LDAP_ALT_USERNAME" value="anotherUser" />
-        <env name="TESTS_ZEND_LDAP_ALT_DN" value="CN=Another User,OU=Sales,DC=example,DC=com" />
-        <env name="TESTS_ZEND_LDAP_ALT_PASSWORD" value="" />
-        <env name="TESTS_ZEND_LDAP_WRITEABLE_SUBTREE" value="OU=Test,OU=Sales,DC=example,DC=com" />
+        <env name="TESTS_ZEND_LDAP_ALT_USERNAME" value="user1" />
+        <env name="TESTS_ZEND_LDAP_ALT_PRINCIPAL_NAME" value="user1@example.com" />
```
Comment (heiglandreas) [inline phpunit.xml.dist]: Renaming this user causes a lot of tests to fail due to having different values being expected in the tests

File: .ci/ldif/user1.example.com.ldif [hunk 3]
```diff
@@ -0,0 +1,5 @@
+dn: uid=user1,dc=example,dc=com
```
Comment (heiglandreas) [inline .ci/ldif/user1.example.com.ldif]: This user is in some of the tests expected to be named ```anotherUser``` and causes these tests to fail

File: .ci/ldif/test.example.com.ldif [hunk 4]
```diff
@@ -0,0 +1,3 @@
+dn: ou=test,dc=example,dc=com
+objectClass: organizationalUnit
```
Comment (heiglandreas) [inline .ci/ldif/test.example.com.ldif]: There should be a base of ```dc=Sales, dc=example,dc=com``` to have the possibility to check for subtree-searches.

File: phpunit.xml.dist (hunk 1 above)
Comment (Maks3w) [inline phpunit.xml.dist]: 1. I change it for match the setup instructions
2. My idea is to replicate a test environment with production needs and features
3. Test suite is not expected to be executed in arbitrary custom directory installations
4. I don't see this values should be customized this is most a DRY

There is no failure in tests. May you are using an old phpunit.xml version

File: test/BindTest.php [hunk 5]
```diff
@@ -64,7 +64,7 @@ public function setUp()
         if (getenv('TESTS_ZEND_LDAP_ALT_USERNAME')) {
             $this->altUsername = getenv('TESTS_ZEND_LDAP_ALT_USERNAME');
         }
-        $this->principalName = getenv('TESTS_ZEND_LDAP_PRINCIPAL_NAME');
+        $this->altPrincipalName = getenv('TESTS_ZEND_LDAP_ALT_PRINCIPAL_NAME');
```
Comment (Maks3w) [inline test/BindTest.php]: @heiglandreas I had to change this because the Manager use is not searchable with the search filter

File: phpunit.xml.dist (hunk 1 above)
Comment (heiglandreas) [inline phpunit.xml.dist]: I've checked out your branch and did the following:

    vagrant up
    export TESTS_ZEND_LDAP_ONLINE_ENABLED=true
    phpunit

A lot of the tests that are skipped without the ```TESTS_ZEND_LDAP_ONLINE_ENABLED=true``` are now failing. Perhaps I did something wrong. I'll dig into it tonight (tight schedule today....)

Comment (Maks3w) [inline phpunit.xml.dist]: Did you remove phpunit. xml?

Comment (heiglandreas) [inline phpunit.xml.dist]: Removed ```phpunit.xml``` and just tried it with ```phpunit -c phpunit.xml.dist```.

Same result. :frowning:

Comment (Maks3w) [inline phpunit.xml.dist]: Your case is very weird. Travis-cI works fine same on my pc and I guess weierophinney too.

Comment (heiglandreas) [inline phpunit.xml.dist]: Works now perfectly. The VM had decided to die suddenly without a notice. Therefore no LDAP-Server was available. After booting up vagrant AGAIN everything works. Sorry for the disturbance. 

So  after remove the ```phpunit.xml``` that's fine and ready to merge from my side!

Thanks for the effort!

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -75,6 +75,18 @@ To do so:
  -  Edit `phpunit.xml` to enable any specific functionality you
     want to test, as well as to provide test values to utilize.
 
+### Execute tests against a real LDAP server
```
Comment (Maks3w) [inline CONTRIBUTING.md]: @weierophinney @heiglandreas Please review the instructions.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -75,6 +75,18 @@ To do so:
...
+
+This projects provides a [Vagrant](https://www.vagrantup.com/) virtual machine definition with an OpenLDAP server
+installed and configured.
+
+For to execute the tests against this server do the following steps (assume you have Vagrant > 1.6 installed)
+
+1. Run the virtual machine with the LDAP server: `vagrant up`
+2. Copy `phpunit.xml.dist` file to `phpunit.xml`
```
Comment (heiglandreas) [inline CONTRIBUTING.md]: I would skip this part and add it at the end of the section like "If you want to alter the configuration for the unit-tests you might want to duplicate the existing phpunit.xml.dist-file to phpunit.xml."
